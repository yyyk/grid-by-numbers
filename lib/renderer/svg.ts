import { degToRad } from '../maths/helpers/degToRad'
import {
  getIdentityMatrix,
  rotate as mat3Rotate,
  scale as mat3Scale,
  translate as mat3Translate,
  Mat3
} from '../maths/mat3'
import { transformByMat3, Vec3 } from '../maths/vec3'
import { Vec2 } from './../maths/vec2'

export function getTransformMatrix(
  size: Vec2,
  position: Vec2,
  rotateRadian: number,
  scale: Vec2
): Mat3 {
  let matrix: Mat3 = getIdentityMatrix()
  const x: number = position[0] + size[0] / 2
  const y: number = position[1] + size[1] / 2
  // position
  matrix = mat3Translate(matrix, [x, y])
  // translate
  // matrix = mat3Translate(matrix, translate)
  // rotate
  matrix = mat3Rotate(matrix, rotateRadian)
  // scale
  matrix = mat3Scale(matrix, scale)
  return matrix
}

export function getTransformMatrixString(
  size: Vec2,
  position: Vec2,
  rotateRadian: number,
  scale: Vec2
): string {
  let matrix: Mat3 = getIdentityMatrix()
  const x: number = position[0] + size[0] / 2
  const y: number = position[1] + size[1] / 2
  // position
  matrix = mat3Translate(matrix, [x, y])
  // translate
  // matrix = mat3Translate(matrix, translate)
  // rotate
  matrix = mat3Rotate(matrix, rotateRadian)
  // scale
  matrix = mat3Scale(matrix, scale)
  // matrix(1,0,0,1,tx,ty)
  return `matrix(${matrix
    .filter((_, i: number) => i !== 2 && i !== 5 && i !== 8)
    .join(', ')})`
}

export interface Group {
  transform: string
}

export function getGroup(
  position: Vec2 = [0, 0],
  rotateDegree = 0,
  scale: Vec2 = [1, 1]
): Group {
  const rotateRadian: number = degToRad(rotateDegree)
  const matrix: string = getTransformMatrixString(
    [0, 0],
    position,
    rotateRadian,
    scale
  )
  return {
    transform: matrix
  }
}

export interface Rect {
  x: number
  y: number
  width: number
  height: number
  transform: string
}

export function getRect(
  size: Vec2,
  position: Vec2,
  rotateDegree = 0,
  scale: Vec2 = [1, 1]
): Rect {
  const halfWidth: number = size[0] / 2
  const halfHeight: number = size[1] / 2
  const rotateRadian: number = degToRad(rotateDegree)
  const matrix: string = getTransformMatrixString(
    size,
    position,
    rotateRadian,
    scale
  )
  return {
    x: -halfWidth,
    y: -halfHeight,
    width: size[0],
    height: size[1],
    transform: matrix
  }
}

export interface Path {
  d: string
}

export function getRectPath(
  size: Vec2,
  position: Vec2,
  rotateDegree = 0,
  scale: Vec2 = [1, 1]
): Path {
  const halfWidth: number = size[0] / 2
  const halfHeight: number = size[1] / 2
  let positions: Vec3[] = [
    [-halfWidth, -halfHeight, 1],
    [halfWidth, -halfHeight, 1],
    [halfWidth, halfHeight, 1],
    [-halfWidth, halfHeight, 1]
  ]
  const rotateRadian: number = degToRad(rotateDegree)
  const matrix: Mat3 = getTransformMatrix(size, position, rotateRadian, scale)
  positions = positions.map((pos) => transformByMat3(pos, matrix))
  const path = `M${positions[0][0].toFixed(2)} ${positions[0][1].toFixed(
    2
  )} L${positions[1][0].toFixed(2)} ${positions[1][1].toFixed(
    2
  )} L${positions[2][0].toFixed(2)} ${positions[2][1].toFixed(
    2
  )} L${positions[3][0].toFixed(2)} ${positions[3][1].toFixed(2)} Z`
  return {
    d: path
  }
}
