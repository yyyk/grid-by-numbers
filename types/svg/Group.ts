export interface GTagAttr {
  id: string
  tag: string
  positionX?: number
  positionY?: number
  rotate?: number
  scaleX?: number
  scaleY?: number
  children?: any[] // TODO:
}
