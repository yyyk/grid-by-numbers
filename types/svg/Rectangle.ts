export interface RectTagAttr {
  id: string
  tag: string
  width?: number
  height?: number
  positionX?: number
  positionY?: number
  rotate?: number
  scaleX?: number
  scaleY?: number
}
