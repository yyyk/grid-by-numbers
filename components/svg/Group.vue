<template>
  <g :id="attributes.id" :transform="computedTag.transform">
    <template v-if="attributes.children && attributes.children.length > 0">
      <template v-for="child in attributes.children">
        <Group
          v-if="child.tag === 'group'"
          :key="child.id"
          :attributes="child"
        />
        <Rectangle
          v-else-if="child.tag === 'rectangle'"
          :key="child.id"
          :attributes="child"
        />
      </template>
    </template>
  </g>
</template>

<script lang="ts">
import Vue, { PropOptions } from 'vue'
import { getGroup } from '../../lib/renderer/svg'
import { Vec2 } from '../../lib/maths/vec2'
import Rectangle from './Rectangle.vue'
import { GTagAttr } from '@/types/svg/Group'

export default Vue.extend({
  name: 'Group',
  components: {
    Rectangle
  },
  props: {
    attributes: {
      type: Object,
      required: true
    } as PropOptions<GTagAttr>
  },
  computed: {
    computedTag() {
      const attr = this.attributes
      const position: Vec2 = [attr.positionX || 0, attr.positionY || 0]
      const scale: Vec2 = [
        attr.scaleX == null ? 1 : attr.scaleX,
        attr.scaleY == null ? 1 : attr.scaleY
      ]
      return getGroup(position, attr.rotate || 0, scale)
    }
  }
})
</script>
