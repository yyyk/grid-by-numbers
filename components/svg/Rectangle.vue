<template>
  <rect
    :id="attributes.id"
    :x="computedTag.x"
    :y="computedTag.y"
    :width="computedTag.width"
    :height="computedTag.height"
    :transform="computedTag.transform"
    fill="none"
    stroke="#000"
    stroke-width="1"
    vector-effect="non-scaling-stroke"
  />
</template>

<script lang="ts">
import Vue, { PropOptions } from 'vue'
import { getRect } from '../../lib/renderer/svg'
import { Vec2 } from '../../lib/maths/vec2'
import { RectTagAttr } from '@/types/svg/Rectangle'

export default Vue.extend({
  name: 'Rectangle',
  props: {
    attributes: {
      type: Object,
      required: true
    } as PropOptions<RectTagAttr>
  },
  computed: {
    computedTag() {
      const attr = this.attributes
      const size: Vec2 = [attr.width || 0, attr.height || 0]
      const position: Vec2 = [attr.positionX || 0, attr.positionY || 0]
      const scale: Vec2 = [
        attr.scaleX == null ? 1 : attr.scaleX,
        attr.scaleY == null ? 1 : attr.scaleY
      ]
      return getRect(size, position, attr.rotate || 0, scale)
    }
  }
})
</script>
