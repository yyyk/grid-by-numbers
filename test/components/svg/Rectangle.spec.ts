import { mount, createLocalVue } from '@vue/test-utils'
import Rectangle from '@/components/svg/Rectangle.vue'
import { RectTagAttr } from '@/types/svg/Rectangle'

const localVue = createLocalVue()

describe('Renderer.vue', () => {
  const mountFunction = (options = {}) => {
    return mount(Rectangle, {
      localVue,
      ...options
    })
  }

  it('renders correctly', () => {
    const wrapper = mountFunction({
      propsData: {
        attributes: {
          id: '0',
          tag: 'rectangle'
        } as RectTagAttr
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('exists', () => {
    const wrapper = mountFunction({
      propsData: {
        attributes: {
          id: '0',
          tag: 'rectangle'
        } as RectTagAttr
      }
    })
    expect(wrapper.exists()).toBe(true)
  })

  // TODO: computed
})
