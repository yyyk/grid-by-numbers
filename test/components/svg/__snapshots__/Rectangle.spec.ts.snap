// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer.vue renders correctly 1`] = `"<rect id=\\"0\\" x=\\"0\\" y=\\"0\\" width=\\"0\\" height=\\"0\\" transform=\\"matrix(1, 0, 0, 1, 0, 0)\\" fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"1\\" vector-effect=\\"non-scaling-stroke\\"></rect>"`;
